pkg/gadgets/traceloop/tracer/bpf/traceloop.bpf.c

```diff
@@ -0,0 +1,454 @@
+	struct syscall_def_t *syscall_def;
+	struct task_struct *task;
+	long ret = ctx->args[1];
+	struct pt_regs *args;
+	void *perf_buffer;
+	u64 mntns_id;
+	int i, r;
+	u64 nr;
+
+	r = bpf_map_update_elem(&regs_map, &pid, &empty, BPF_NOEXIST);
+	if (r) {
+		bpf_printk("exit: there should not be any pt_regs for key %lu: %d\n", pid, r);
+
+		return 0;
+	}
+
+	args = bpf_map_lookup_elem(&regs_map, &pid);
+	if (!args) {
+		bpf_printk("exit: there should be a pt_regs for key %lu\n", pid);
+
+		goto end;
+	}
+
+	bpf_probe_read(args, sizeof(*args), (void*) ctx->args[0]);
+	nr = syscall_get_nr(args);
+	/* TODO Why this can occur? */
+	if (nr == -1)
+		goto end;
+
+	struct syscall_event_t sc = {
+		.cpu = bpf_get_smp_processor_id(),
+		.pid = pid >> 32,
+		.typ = SYSCALL_EVENT_TYPE_EXIT,
+		.id = nr,
+	};
+	sc.args[0] = ret;
+
+	syscall_def = bpf_map_lookup_elem(&syscalls, &nr);
+	if (syscall_def == NULL)
+		syscall_def = &default_definition;
+
+	task = (struct task_struct*)bpf_get_current_task();
+	mntns_id = (u64) BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);
+
+	perf_buffer = bpf_map_lookup_elem(&map_of_perf_buffers, &mntns_id);
+	if (!perf_buffer)
+		goto end;
+
+	remembered = bpf_map_lookup_elem(&probe_at_sys_exit, &pid);
+	if (!remembered)
+		goto end;
+
+	/*
+	 * This ensures all events (enter, exit and cont) related to a given
+	 * syscall have the same timestamp.
+	 */
+	sc.timestamp = remembered->timestamp;
+
+	for (i = 0; i < SYSCALL_ARGS; i++) {
+		__u64 arg_len = syscall_def->args_len[i];
+
+		if (!arg_len || !(arg_len & PARAM_PROBE_AT_EXIT_MASK))
+			goto end_loop;
+
+		bool null_terminated = false;
+		struct syscall_event_cont_t sc_cont = {
+			.timestamp = remembered->timestamp,
+			.index = i,
+			.failed = false,
+		};
+
+		arg_len &= ~PARAM_PROBE_AT_EXIT_MASK;
+
+		if (arg_len == USE_RET_AS_PARAM_LENGTH) {
+			if ((signed long) ret < 0)
+				arg_len = 0;
+			else
+				arg_len = ret;
+		} else if (arg_len == USE_NULL_BYTE_LENGTH) {
+			null_terminated = true;
+			arg_len = 0;
+		} else if (arg_len >= USE_ARG_INDEX_AS_PARAM_LENGTH) {
+			__u64 idx = arg_len & USE_ARG_INDEX_AS_PARAM_LENGTH_MASK;
+			if (idx < SYSCALL_ARGS)
+				arg_len = remembered->args[idx];
+			else
+				arg_len = PARAM_LEN;
+		}
+
+		if (arg_len > sizeof(sc_cont.param))
+			arg_len = sizeof(sc_cont.param);
+
+		if (null_terminated)
+			sc_cont.length = USE_NULL_BYTE_LENGTH;
+		else
+			sc_cont.length = arg_len;
+
+		// On Linux 4.14.137+, calling bpf_probe_read() with a variable size causes:
+		// "invalid stack type R1 off=-304 access_size=0"
+		// This is fixed on newer kernels.
+		//
+		// I know arg_len is not a volatile but that stops the compiler from
+		// optimising the ifs into one bpf_probe_read call with a variable size.
+		/* enforces zero-termination in sc_cont.param even if the string is larger */
+		if (arg_len == 0
+			&& null_terminated
+			&& bpf_probe_read_user_str(sc_cont.param, PARAM_LEN, (void *)(remembered->args[i])) < 0)
```

Comment:
I think `bpf_probe_read_user_str` should be done inside the 'then' block instead of inside the `if` condition because if `bpf_probe_read_user_str` returns success (0), you don't want to evaluate the other `else` condition.